File: implementations/python/src/esque/elasticsearch/lock.py
from __future__ import annotations

import logging
import threading
import time

from esque.elasticsearch.operations import RestClientOperations

log = logging.getLogger(__name__)

_IDLE_BETWEEN_TRIES = 0.1  # seconds — mirrors ElasticsearchDocumentLock.DEFAULT_IDLE_BETWEEN_TRIES


class ElasticsearchDocumentLock:
    def __init__(self, operations: RestClientOperations) -> None:
        self._operations = operations
        self._delegate = threading.RLock()

    def try_lock(self, timeout_minutes: int) -> bool:
        deadline = time.monotonic() + timeout_minutes * 60
        remaining = max(0.0, deadline - time.monotonic())

        if not self._delegate.acquire(blocking=True, timeout=remaining):
            return False

        while True:
            try:
                if self._do_lock():
                    return True
            except Exception:
                self._delegate.release()
                raise

            if time.monotonic() >= deadline:
                self._delegate.release()
                return False

            time.sleep(_IDLE_BETWEEN_TRIES)

    def unlock(self) -> None:
        try:
            self._operations.delete_lock_record()
        except Exception as e:
            raise RuntimeError("Failed to release mutex") from e
        finally:
            self._delegate.release()

    def _do_lock(self) -> bool:
        try:
            self._operations.create_lock_record()
            return True
        except Exception as e:
            # TODO: differentiate ConflictError (lock exists) from other failures
            log.debug("Failed to acquire lock: %s", e)
            return False

File: implementations/typescript/src/elasticsearch/lock.ts
import { setTimeout as sleep } from "node:timers/promises";
import type { RestClientOperations } from "./operations.js";

const IDLE_BETWEEN_TRIES_MS = 100;

// Thrown by unlock() when the lock is not held — expected during Esque.close() after a clean run.
export class LockNotHeldError extends Error {}

// Distributed lock via ES op_type=create. The JVM wraps this in a local ReentrantLock for
// thread safety; Node is single-threaded, so a held-flag suffices — it exists to make
// unlock() without tryLock() a detectable error, and to avoid deleting another process's
// lock document from a process that never acquired it.
export class ElasticsearchDocumentLock {
  private readonly operations: RestClientOperations;
  private held = false;

  constructor(operations: RestClientOperations) {
    this.operations = operations;
  }

  async tryLock(timeoutMinutes: number): Promise<boolean> {
    const deadline = Date.now() + timeoutMinutes * 60_000;
    for (;;) {
      if (await this.doLock()) {
        this.held = true;
        return true;
      }
      if (Date.now() >= deadline) {
        return false;
      }
      await sleep(IDLE_BETWEEN_TRIES_MS);
    }
  }

  async unlock(): Promise<void> {
    if (!this.held) {
      throw new LockNotHeldError("cannot release un-acquired lock");
    }
    this.held = false;
    try {
      await this.operations.deleteLockRecord();
    } catch (error) {
      throw new Error("Failed to release mutex", { cause: error });
    }
  }

  private async doLock(): Promise<boolean> {
    try {
      await this.operations.createLockRecord();
      return true;
    } catch {
      // TODO: differentiate ConflictError (lock exists) from other failures
      return false;
    }
  }
}
